### ISSUE TITLE
[Llama-Engine] Design: parity test harness — llama-engine vs llama-server (quality) and rpc-server (compute)

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/hydra_vortex#527]
## Summary

Design for a **llama-engine parity test harness** on the engine side: a reusable,
deterministic driver that proves our `llama-engine` implementation meets the
**quality** and **performance** of the two reference binaries it replaces on the
wire —

- **`llama-server`** — the reference for OpenAI-compatible HTTP surface, sampling
  quality, and single-GPU prefill/decode throughput.
- **`rpc-server`** (`tools/rpc/rpc-server.cpp`) — the reference for serving a GPU
  as a remote ggml-RPC compute backend (the dimension the engine's embedded RPC
  already self-measures against, see `llama-engine.cpp:635-651`).

**"Test harness first"** — before further llama-engine feature work, stand up the
harness so every engine change can be validated as a candidate against both
references on the same model/GPU/context. This follows the #470 design model:
#470 closed the *runtime* safety gap (validate-before-KV-lands) inside the engine;
this harness closes the *verification* gap — it is the external gate that proves
the engine actually behaves like the reference it replaced.

## References

- #470 — merged P/D `DECODE` design (two-gate validation: Gate A sync header
  check, Gate B post-load identity check). The harness must exercise both gates.
- #306 — existing `tests/bench` suite (advisory, Coordinator-level). Reuse its
  percentile/compare methodology; this harness is engine-level (no Coordinator).
- #518 — Python→xUnit+Aspire test migration. This harness is a python driver for
  now; keep the scenario list portable so it can be re-authored later.
- `tests/bench/ab_engine.py` — existing A/B functional driver (engine RPC 0x40-46
  vs legacy HTTP on the *same* binary). Extend/reuse its framing helpers; this
  design compares *separate binaries* against each other.
- Fork unit tests `tests/test-hydra-*.cpp` (configure-tier, state-chunk, rpc-bind,
  checkpoint-policy, model-identity) — fast unit gates already exist; this harness
  is the end-to-end + perf complement.
- `docs/llama-bench-guide.md` + `docs/eval-tests.md` — perf invocation + report/pass
  criteria formats to reuse.

## What we are asserting

| Dimension | Reference | Assertion |
|---|---|---|
| HTTP quality | `llama-server` | Same request → identical token stream + OpenAI schema (SSE chunk shape, `reasoning_content`, `finish_reason`, `usage`, `timings`, error codes) |
| HTTP performance | `llama-server` | pp/tg tok/s and TTFT/TPOT within tolerance (same model, GPU, ctx, cache types) |
| RPC compute quality | `rpc-server` | Remote head computes the same graph → bit-identical result tensors |
| RPC compute performance | `rpc-server` | Head-side prefill/decode tok/s within tolerance when the peer is engine-embedded vs standalone |
| Engine control surface | protocol spec | `0x40`-`0x46` semantics per `specs/rpc-protocol.md`, incl. #470 Gate A/B behavior |
| Determinism | — | temp=0 + fixed seed + fixed shapes → identical token IDs across processes (the base for all equality checks) |

## Harness architecture

New Python package **`tests/engine_harness/`** in hydra_vortex (sibling of
`tests/bench`), driven against fork build artifacts:

```
tests/engine_harness/
  run.py                  CLI: --binary-dir, --model, --vs {server,rpc}, --suite {quality,perf,all}
  drivers.py              spawn/stop llama-server, llama-engine, rpc-server; wait /health; cleanup
  rpc_client.py           reuse ab_engine.py framing (extract shared wire helpers)
  sse.py                  SSE chunk parser + strict OpenAI-schema validator
  reference.py            run reference binary, capture tokens/timings/schema
  candidate.py            run llama-engine, same capture
  quality.py              Q1-Q9 scenarios
  perf.py                 P1-P5 scenarios
  rpc_parity.py           head-mode client for engine-embedded vs standalone rpc-server
  report.py               markdown + JSON report (eval-tests.md style)
```

Determinism harness: greedy (`temp=0`), fixed `seed`, fixed `n_ctx`, fixed
`--cache-type-k/v`, warmup discard. Token equality = exact match of sampled token
IDs (not just text), because text equality hides reasoning/hidden-token drift.

## Quality scenarios

- **Q1** HTTP chat completion, non-stream — engine vs server: schema + content equality.
- **Q2** HTTP chat completion, SSE stream — chunk-by-chunk equality + SSE grammar.
- **Q3** `reasoning_content`/`content` split (both `--reasoning on`) — engine vs server.
- **Q4** `0x42` PREFILL → `n_past` + KV blob; verify the blob round-trips through the
  same engine's `0x43` decode (prefill-compat, mirrors #470).
- **Q5** `0x43` DECODE with matching `kv_metadata` (Gate A pass) → SSE results,
  restored-logits path exercised.
- **Q6** `0x43` DECODE with wrong-model `kv_metadata` → **Gate A reject**, no KV
  written, expected error code.
- **Q7** `0x43` Gate B: unknown alias fallback → `RESIDENT_IDENTITY_MISMATCH`, no
  KV write (the #469/#470 safety property, checked externally).
- **Q8** Model swap via `0x42` `model`/`hydra_config` (T3) → identity in response;
  then decode on the swapped model; assert no crash + correct `model_alias`.
- **Q9** Error paths: `BUSY`, `BAD_REQUEST`, malformed payloads, `HDR_HASH_MISMATCH`,
  `SEGMENT_TABLE_INVALID`, `SEGMENT_HASH_MISMATCH`.

## Performance scenarios

- **P1/P2** Prefill (`pp512/4096/32768`) + decode (`tg64/128`) via **HTTP** —
  engine vs llama-server, same model/GPU/ctx. Report P50 tok/s, `compare.py`
  tolerance gate.
- **P3** TTFT/TPOT P50/P95/P99, single + `--concurrency N` — engine vs server.
- **P4** RPC-backend parity: `llama-bench --rpc <engine-rpc-port>` vs
  `--rpc <rpc-server-port>` with identical `-ot` placement → pp/tg tok/s within
  tolerance. This is the direct check for the known pool/peer-mode concern
  (llama-engine.cpp:635-651).
- **P5** P/D end-to-end (RTX prefill → KV → P100 `0x43` decode) vs direct-server
  decode baseline, using the eval-tests.md P/D pass criteria.

## Exit criteria / report

- Q1-Q9 pass (schema + token-ID equality) and P1-P5 within tolerance on a pinned
  model + binary set → **PARITY PASS**.
- Report at `tests/results/engine-parity-{ts}.md` + `.json` (advisory at first,
  like tests/bench), single-pager of the eval-tests.md report format.
- Any drift between engine and reference is filed as a `review-finding` issue
  before the offending change merges (engine parity becomes the per-change gate).

## CI integration (proposal)

- New `engine-parity.yml` on the fork or hydra_vortex: `workflow_dispatch`
  (hydra-build.yml pattern), runs on the local sm86-sm120 runner with a GPU.
- Model: pinned Qwopus3.6-35B-A3B (Q3_K-mini) for perf; a smaller smoke model for
  fast Q-suite runs.
- Advisory first; promote to blocking for the Llama-Engine milestone once stable
  (a run must fit in ~15 min on the local runner).

## Sub-phases (for PROJECT_STATUS.md)

| Phase | What | Exit |
|---|---|---|
| 0 | Extract shared wire helpers from `ab_engine.py`; driver + determinism harness | Same token IDs across 2 runs |
| 1 | Q1-Q3 HTTP/schema parity vs llama-server | PASS on pinned model |
| 2 | Q4-Q9 RPC control surface + #470 gates | Gate A/B reject paths verified |
| 3 | P1-P3 HTTP perf parity | Within tolerance vs server |
| 4 | P4 RPC-backend parity vs rpc-server | Within tolerance (validates peer_mode/pool sizing) |
| 5 | P5 P/D end-to-end + CI workflow + report | Advisory gate in CI |

## Open questions

1. Determinism tolerance: is token-ID equality stable enough across process
   restarts on this stack, or do we need a documented "repro lists" warmup/skip
   (CUDA nondeterminism)? Decide before Phase 1 gates on exact equality.
2. Perf tolerance value (±5% vs ±10%)? Start ±10%, tighten once baseline noise is
   measured.
3. Harness home: hydra_vortex (recommended, alongside tests/bench) vs inside the
   fork's `tests/`. If fork-side, the fork's "no AI PRs" policy applies only to
   upstream; this is a private-fork/private-repo concern.
4. `0x43` SSE result delivery is via HTTP `/v1/decode/{id}` (protocol v3) — Q5
   needs the poll loop (202→200/400). Confirm the harness asserts the 400-abort
   rule (never proxy) as part of Q7.
5. Does P4 need a dedicated head-mode client, or is `llama-bench --rpc` with
   `-ot` placement sufficient as the reference compute workload?

## Related

- #514 (2-4 tok/s after swap) and #508/#504/#499 — the class of regressions this
  harness exists to catch before they ship.
- #397 (llama-engine v4 design) — milestone tracking parent.
- #306 / #518 — adjacent test-suite work to stay coordinated with.
